_drafts/restless-restfull.md

```diff
@@ -0,0 +1,776 @@
+talkative intern, there were much wiser people guiding me.
+
+Some of the projects were restricted to internal proprietary protocols,
+others were using RTMP (thank God that is over), but we faced some opportunity
+to do what all the other cool kids where doing, shiny new REST API's.
+At that time it as so well established that WS-\* stuff and SOAP where pure
+evil that I didn't even bother to look that up too much and delve directly
+on REST and its blazing glory.
+
+When you have no idea what you are doing because something is new
+to you the first instinct is too come up with a set of rules to
+avoid stupid mistakes and to avoid handling with context that
+you can't understand yet, this fits with the
+[Dreyfus Model Of Skill Acquisition](https://en.wikipedia.org/wiki/Dreyfus_model_of_skill_acquisition).
+
+Searching for guidance on this aspect we found several guidelines
+that helped us build our own. At this stage there was just
+some web searching. The guidelines were sound, that usual stuff
+of handling collections of stuff and individual stuff on a
+hierarchical way using URL's, the traditional dogs example:
+
+Getting all dogs:
+
+```
+GET
+/myapi/dogs
+```
+
+Getting individual dog:
+
+```
+GET
+/myapi/dogs/someid
+```
+
+Using verbs on URL names mean you got everything wrong and was a bad bad
+developer, you where not thinking on the RESTful way and should feel
+bad about yourself. There was a **LOT** of focus on the CRUD thing too,
+on how the HTTP methods should be used correctly. Not that these are
+bad ideas, it is a good idea to have a uniform system where the intentions
+are clear. No one wants to use the **DELETE** method to get some data, etc.
+But I started to get frustrated on how almost all material that I was reading
+focused way too much on URL's naming and HTTP verbs.
+
+They always explain the collection of resources thing, and them how to related
+resources, like the dog has an owner, how to express that ? Again, not that it
+is a bad idea to think on how to properly handle this kind of relationship, but to
+be honest it was not one of my biggest problems at the time. Actually if
+you squint you will see that discussing this is the same as discussing how
+to organize data on a database.
+
+Since URL's are hierarchical you will have similar problems to when you
+represent sets of information (or resources) that have relationships between
+them using a document oriented database (hierarchical). You will need to think
+if you always send the resource with its related resources embedded, or return
+links, there will be tradeoff's with each approach, etc. There is a good space
+for discussion on this area, and it is good discussion, depending on the
+kind of system you are developing different sets of tradeoff's must be made.
+
+A great article that helped me understand the solution space better is
+[A Relational Model of Data for Large Shared Data Banks](https://www.seas.upenn.edu/~zives/03f/cis550/codd.pdf).
```

Comment:
@ryukinix My english is 100% informal learned from playing video games and watching TV, so it is considerably crappy =P (I have 0% notion of grammar rules), thanks for helping with that ;-)

Reply:
I'm not so different. I also have really serious problems about bad grammar ([ex]). For all matters, I learned a lot of the vocabulary and other stuff just playing video games and talking with random people through internet. But... well, sometimes I like study grammar because it's painful being misunderstood :v  

[ex]: https://github.com/ryukinix/decorating/issues/11

Reply:
Too lazy to study grammar, specially ones for natural languages. But I do enjoy BNF style grammar descriptions for context free languages =D